substrate/frame/support/src/storage/state_diff_guard.rs

```diff
@@ -0,0 +1,381 @@
+//!
+//! # Example
+//!
+//! Use the `StateDiffGuard` in a migration:
+//!
+//! ```rust
+//! 
+//! use frame_support::storage::generator::{StorageMap, StorageValue};
+//! use frame_support::storage::StateDiffGuard;
+//! use frame_support::traits::Get;
+//! use sp_io::TestExternalities;
+//!
+//! #[frame_support::pallet]
+//! mod pallet {
+//! 	use frame_support::pallet_prelude::*;
+//!
+//! 	#[pallet::pallet]
+//! 	pub struct Pallet<T>(_);
+//!
+//! 	#[pallet::call]
+//! 	impl<T: Config> Pallet<T> {
+//! 		#[pallet::weight(0)]
+//! 		pub fn set_value(origin: OriginFor<T>, value: u32) -> DispatchResult {
+//! 			<Value<T>>::put(value);
+//! 			Ok(())
+//! 		}
+//! 	}
+//!
+//! 	#[pallet::storage]
+//! 	pub type Value<T> = StorageValue<_, u32>;
+//!     
+//! 	#[pallet::storage]
+//! 	pub type SomeMap<T> = StorageMap<_, Twox64Concat, u32, u32>;
+//!     
+//!     #[pallet::storage]
+//! 	pub type SomeDoubleMap<T> = StorageMap<_, Twox64Concat, u32, u32>;
+//! }
+//!
+//! mod migrations {
+//! 	use super::*;
+//!
+//!     pub struct UncheckedMigrateToV1<T: crate::Config>(sp_std::marker::PhantomData<T>);
+//!
+//!     impl<T: crate::Config> frame_support::traits::UncheckedOnRuntimeUpgrade for UncheckedMigrateToV1<T> {
+//!         fn on_runtime_upgrade() -> frame_support::weights::Weight {
+//! 		   // migration logic here
+//!            Weight::default()
+//!         }
+//!         
+//!         #[cfg(feature = "try-runtime")]
+//!         fn try_on_runtime_upgrade() -> Result<frame_support::weights::Weight, &'static str> {
+//! 		   // migration logic here
+//!           let guard = StateDiffGuard::builder()
+//! 			.must_change(StoragePrefix {
+//! 				pallet_name: SomeMap::<T>::pallet_prefix(),
+//! 				storage_name: SomeMap::<T>::storage_prefix(),
+//! 			})
+//! 			.must_change(StoragePrefix {
+//! 				pallet_name: SomeDoubleMap::<T>::pallet_prefix(),
+//! 				storage_name: SomeDoubleMap::<T>::storage_prefix(),
+//! 			})
+//! 			.build();
+//!
+//!           // try runtime upgrade logic here
+//! 		  let weight = Self::on_runtime_upgrade();
+//!           
+//!           // any other logic
+//! 		  Ok(weight)
+//! 	   }
+//!     }
+//!
+//!     pub type MigrateToV1<T> = frame_support::migrations::VersionedMigration<
+//!     	1,
+//!      	2,
+//!      	UncheckedMigrateToV1<T>,
+//!      	Pallet<T>,
+//! 		<T as frame_system::Config>::DbWeight
+//!     >;
+//! }
+//! ```
+//!
+//! In the example above, the guard will panic if any storage entry that doesn't match `SomeMap` or `SomeDoubleMap`
+//! prefixes is changed.
+
+use core::fmt::{Debug, Formatter};
+use sp_std::{
+	collections::{btree_map::BTreeMap, btree_set::BTreeSet},
+	vec::Vec,
+};
+
+use super::storage_prefix;
+
+/// Raw storage key.
+type StorageKey = Vec<u8>;
+/// Raw storage value.
+type StorageValue = Vec<u8>;
+
+/// Storage prefix: pallet name and storage name.
+#[derive(Clone, Copy, Ord, PartialOrd, Eq, PartialEq)]
+pub struct StoragePrefix {
```

Comment:
We should already have something like this struct in frame-support. `trait StorageInstance` or `struct StorageInfo` for example.